### PR TITLE
fix(security): Fix Path Traversal in model generation

### PR DESCRIPTION
**Pull request by anudeeps0306**
## Description

This PR hardens file path handling to reduce path manipulation risks caused by manual file extension concatenation and fragile filename construction. It ensures all file accesses use safer path joining and validation, and avoids relying solely on `secure_filename` in cases where unexpected payloads or poisoned DB values could lead to incorrect paths.

## Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation update

## How Has This Been Tested?

- [x] Existing tests pass
- [ ] New tests added
- [x] Manual testing

Manual testing included:
- Uploading files with normal and edge-case filenames
- Verifying generated paths are correct and constrained to the expected directory
- Verifying no regressions in existing file read/write flows

## Screenshots (if applicable)

N/A

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review
- [ ] I have added/updated documentation as needed
- [x] My changes generate no new warnings
- [x] Tests pass locally

### CODE REVIEW

**Review by ivantha — CHANGES_REQUESTED** [review]
Please rebase

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR is titled as a targeted path-traversal security fix but is in reality a near-complete rewrite of the entire TensorMap stack (Flask → FastAPI, CRA → Vite/Tailwind, server rename, UUID migration, new Project API). The security improvements are real but **incomplete**: `secure_filename` is applied in two new locations yet neither location follows up with an absolute-path containment check, leaving partial traversal exposure. More critically, the ZIP extraction in `data_upload.py` is vulnerable to a ZIP Slip attack (arbitrary file write via crafted archives), and `_get_file_path`-style helpers throughout the backend construct paths from raw DB values without sanitization. A data-destroying `TRUNCATE CASCADE` inside a production Alembic migration is a high-severity operational risk that must be addressed before merging.

File: tensormap-backend/app/services/data_process.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
+import uuid as uuid_pkg
+from typing import Any
+
+import pandas as pd
+from sqlmodel import Session, select
+from werkzeug.utils import secure_filename
+
+from app.config import get_settings
+from app.models import DataFile, DataProcess, ImageProperties
+from app.shared.logging_config import get_logger
```
Comment (ivantha) [inline tensormap-backend/app/services/data_process.py]: **[MEDIUM]** Duplicated File-Path Construction Logic Across Four Services

Identical or near-identical path construction logic exists in four separate places: `data_process._get_file_path`, `code_generation._file_location`, `model_run._helper_generate_file_location`, and inline in `data_upload`. Each independently applies (or omits) sanitization, making it impossible to fix security issues in one place and have them apply everywhere.

**Suggestion:**
```
Create a single shared utility in `app/shared/file_utils.py`:
```python
def get_file_path(upload_folder: str, file_name: str, file_type: str) -> str:
    # sanitize, join, and check containment
    ...
```
Import and call it from all four service files, removing the duplicated implementations.
```

File: tensormap-backend/app/services/model_run.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+
+def _model_result(message: str, test: int) -> None:
+    import asyncio
+
+    data = {"message": message, "test": test}
+    try:
+        loop = asyncio.get_event_loop()
+        if loop.is_running():
+            asyncio.ensure_future(sio.emit(SOCKETIO_LISTENER, data, namespace=SOCKETIO_DL_NAMESPACE))
```
Comment (ivantha) [inline tensormap-backend/app/services/model_run.py]: **[MEDIUM]** asyncio.get_event_loop() Deprecated in Python 3.10+

`asyncio.get_event_loop()` is deprecated since Python 3.10 and emits `DeprecationWarning` on Python 3.12 (the target version). In threads without a running event loop it may raise `RuntimeError`.

```python
loop = asyncio.get_event_loop()
if loop.is_running():
    asyncio.ensure_future(sio.emit(...))
```

**Suggestion:**
```
```python
try:
    loop = asyncio.get_running_loop()
    loop.call_soon_threadsafe(
        asyncio.ensure_future,
        sio.emit(SOCKETIO_LISTENER, data, namespace=SOCKETIO_DL_NAMESPACE)
    )
except RuntimeError:
    asyncio.run(sio.emit(SOCKETIO_LISTENER, data, namespace=SOCKETIO_DL_NAMESPACE))
```
```

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR is presented as a narrow security fix for path traversal vulnerabilities, but it is actually a **complete architectural rewrite**: Flask → FastAPI backend, Semantic UI → Tailwind/shadcn frontend, directory renames, UUID primary keys, new Projects feature, and changed ports. The misleading title and description are themselves a significant issue for reviewers. While the security intent is valid, the implementation is incomplete — `secure_filename` is applied inconsistently (model name saved to DB is unsanitized, file is saved with sanitized name), a Zip Slip vulnerability remains unaddressed in archive extraction, and path construction still uses string concatenation without boundary checks. Most critically, a migration runs a destructive `TRUNCATE … CASCADE` on all tables with auto-migration enabled on startup, which will silently erase all data in any connected database.

**Suggested Version Bump:** `major`

### ⚠️ Breaking Changes

- **[CRITICAL]** Data-Destructive Migration: All Tables Are TRUNCATED: **Previous behavior**: Migration `a1b2c3d4e5f6` (UUID primary keys) runs `TRUNCATE ... CASCADE` before altering column types, destroying all existing data.

**New behavior**: On upgrade, all rows in `model_results`, `model_configs`, `model_basic`, `image_properties`, `data_process`, `data_file`, and `project` are irreversibly deleted.

**Who is affected**: Any environment with existing data (staging, production, any developer with a populated local database).

**Migration path**: There is none — the migration comment says `(dev environment, data loss acceptable)`, but this is committed to a shared migration history that will run on all environments including production if auto-migration is enabled (which it is, per `app/main.py`).
  > **Suggestion:** Replace the destructive `TRUNCATE` with a proper data migration that backfills UUIDs:

```sql
-- Instead of TRUNCATE, use:
ALTER TABLE data_file ADD COLUMN id_new UUID DEFAULT uuid_generate_v4();
UPDATE data_file SET id_new = uuid_generate_v4();
-- ... update foreign keys in child tables ...
ALTER TABLE data_file DROP COLUMN id;
ALTER TABLE data_file RENAME COLUMN id_new TO id;
```
If data loss is truly acceptable for this project stage, add a prominent warning comment and gate the migration behind an environment flag.
- **[HIGH]** Complete API Endpoint Structure Change With New `/api/v1` Prefix: **Previous behavior**: All REST endpoints were served at root paths, e.g., `POST /addData`, `GET /visualizeData`, `POST /saveConfig`.

**New behavior**: All endpoints are now under `/api/v1/`, e.g., `POST /api/v1/data/upload/file`, `GET /api/v1/data/process/target`. Endpoint naming conventions have also changed significantly.

**Who is affected**: Any client, integration, or script that calls the backend API directly (including the current frontend if not updated simultaneously).

**Migration path**: Update all API consumers to use the new base URL `/api/v1` and the new endpoint paths. The `VITE_API_BASE_URL` env var in the frontend `.env.example` already reflects the new base.
- **[HIGH]** Backend and Frontend Ports Changed: **Previous behavior**: Backend served on port 5000, frontend on port 5173.

**New behavior**: Backend serves on port 4300, frontend on port 3300.

**Who is affected**: All Docker deployments, reverse-proxy configurations (nginx, Caddy, etc.), firewall rules, CI/CD environment variables, and developer local setups.

**Migration path**: Update all network configurations, `VITE_API_BASE_URL` settings, reverse-proxy upstreams, and firewall rules to use the new ports.
- **[MEDIUM]** Docker Compose Service Names Changed: **Previous behavior**: Services were named `database`, `server`, and `client`.

**New behavior**: Services are now named `db`, `backend`, and `frontend`.

**Who is affected**: Any external tooling, CI/CD scripts, or documentation that references services by name (e.g., `docker compose exec server bash`, `docker compose logs client`).

**Migration path**: Update all scripts and tooling that reference the old service names.


### PR Description Issues

- **[HIGH]** PR Description-Diff Mismatch: Described as a Security Fix but Is a Full Rewrite: The PR title is `fix(security): Fix Path Traversal in model generation` and the description frames this as a narrow security hardening. In reality, the diff contains:

- **Complete framework migration**: Flask + Poetry → FastAPI + uv
- **Directory rename**: `tensormap-server` → `tensormap-backend`, `tensormap-client` → `tensormap-frontend`
- **Frontend rewrite**: Semantic UI → Tailwind CSS + shadcn/ui
- **New features**: Projects page, Training page, Project management API (`/api/v1/project`)
- **Breaking API changes**: New `/api/v1` prefix, changed ports (5000→4300, 5173→3300)
- **Database schema migration**: Integer primary keys → UUIDs across all tables
- **Docker-compose refactoring**: Service renames (`server`→`backend`, `client`→`frontend`, `database`→`db`), new volumes
- Deletion of the entire `tensormap-server`, `tensormap-client`, `docs` directories

The actual security fixes (applying `yaml.safe_load`, `secure_filename` usage) represent less than 1% of the changed lines. Reviewers who trust the PR title will miss the architectural changes, breaking API surface, and new feature risks.
  > **Suggestion:** Revise the PR description to accurately reflect the scope:

```markdown
## Description
This PR migrates the backend from Flask to FastAPI (with uv package management) and the frontend from a Semantic UI React app to a Tailwind + shadcn/ui app. It also includes security hardening of file path handling and introduces a new Projects management feature.

### Key changes
- **Backend**: Flask → FastAPI; port 5000 → 4300; new `/api/v1` prefix; UUID primary keys
- **Frontend**: Semantic UI → Tailwind/shadcn; port 5173 → 3300
- **New feature**: Project entity for grouping datasets and models
- **Security**: `yaml.safe_load` in templates; `secure_filename` applied to model names and file uploads

### Breaking changes
- All API endpoints are now under `/api/v1/`
- Backend port changed from 5000 to 4300
- Docker service names have changed
```
- **[MEDIUM]** Missing Breaking Changes Section Despite Numerous Breaking Changes: The PR description's 'Type of Change' checklist does not check '- [ ] Breaking change', even though this PR contains multiple user-facing and integrator-facing breaking changes:

1. API base URL path changed (`/` → `/api/v1/`)
2. Backend port changed (5000 → 4300)
3. Frontend port changed (5173 → 3300)
4. docker-compose service names changed (`server` → `backend`, `client` → `frontend`, `database` → `db`)
5. `docker-compose.yml` loses the `version:` key (compatible only with Compose V2+)
6. All primary keys changed from `INTEGER` to `UUID` (existing data is **truncated** by migration `a1b2c3d4e5f6`)
7. The migration explicitly runs `TRUNCATE ... CASCADE` on all tables: `op.execute("TRUNCATE model_results, model_configs, model_basic, image_properties, data_process, data_file, project CASCADE")`
  > **Suggestion:** Check `- [x] Breaking change` in the PR checklist and add an explicit breaking changes section listing all of the above, especially the **data-destructive migration** which will erase all existing data without warning.

**Comment by ivantha** [comment]
Thanks for the PR! It currently has merge conflicts with main. Could you rebase onto the latest main and push the updated branch? Happy to merge once it's auto-rebasable.

**Comment by ivantha** [comment]
Thank you for the path traversal security fix! Unfortunately this PR touches 100 files (lock file churn) making it unreviewable, and has CI failures and change requests. The path traversal vulnerability in deep_learning.py has been addressed more surgically in PR #300. Please feel free to contribute other security improvements as focused, minimal PRs.
